{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2a442877-e9eb-4567-81fb-fb7b971f705c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#         **** Import Libraries  ****\n",
    "\n",
    "import Bio\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqIO.FastaIO import SimpleFastaParser\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from numpy.random import permutation\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import cm\n",
    "%matplotlib inline\n",
    "plt.style.use('bmh')\n",
    "\n",
    "from scipy.spatial.distance import cdist\n",
    "\n",
    "import sys, csv \n",
    "\n",
    "import re\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "import inspect\n",
    "import seaborn as sns\n",
    "import os\n",
    "\n",
    "from itertools import product\n",
    "from pandas.testing import assert_frame_equal\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "472c8386-b42c-4033-b73c-f4f95c0e8792",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#          **** Function Define    ****\n",
    "\n",
    "\n",
    "# Defining a function to read Fasta files and return a dataframe\n",
    "def read_fasta(file_path, columns) :\n",
    " \n",
    "    with open(file_path) as fasta_file :  \n",
    "        records = [] \n",
    "        for title, sequence in SimpleFastaParser(fasta_file): # SimpleFastaParser Iterate over Fasta records as string tuples. For each record a tuple of two strings is returned, the FASTA title line (without the leading ‘>’ character), and the sequence (with any whitespace removed). \n",
    "            record = []\n",
    "            title_splits=re.findall(r\"[\\w']+\", title) # Data cleaning is needed\n",
    "          \n",
    "            record.append(title_splits[0])  # First values are for ID in ngs Fasta \n",
    "            record.append(len(sequence)) # Second values are sequence lengths\n",
    "            sequence = \"\".join(sequence) #It converts into one line\n",
    "            record.append(sequence) # Third values are sequence reads\n",
    "            records.append(record)\n",
    "    return pd.DataFrame(records, columns = columns) # Returns a dataframe\n",
    "\n",
    "\n",
    "\n",
    "#Defining a function to read FastQ Files\n",
    "def load_fastq(file_path):\n",
    "    \"\"\"Load FASTQ file and return a dictionary of SeqRecord objects indexed by their ID.\"\"\"\n",
    "    return {record.id: record for record in SeqIO.parse(file_path, \"fastq\")}\n",
    "\n",
    "#Defining a function to read and store forward and revers reads in a dataframe \n",
    "def match_reads(forward_path, reverse_path):\n",
    "    \"\"\"Match reads from forward and reverse FASTQ files and create a DataFrame.\"\"\"\n",
    "    # Load reads from both files\n",
    "    forward_reads = load_fastq(forward_path)\n",
    "    reverse_reads = load_fastq(reverse_path)\n",
    "\n",
    "    # mMke data for DataFrame\n",
    "    data = []\n",
    "    for idx, (f_id, f_record) in enumerate(forward_reads.items(), start=1):\n",
    "        r_record = reverse_reads.get(f_id, None)\n",
    "        if r_record:  # Ensure reverse read exists\n",
    "            data.append([idx, str(f_record.seq), str(r_record.seq)])\n",
    "    \n",
    "    # Make a DataFrame to store reads\n",
    "    df = pd.DataFrame(data, columns=['Read Number', 'Forward Read', 'Reverse Read'])\n",
    "    return df\n",
    "\n",
    "def reverse_complement(df, column_name):\n",
    "    \"\"\"Reverse complement sequences in a specified column of a DataFrame.\"\"\"\n",
    "    df[column_name] = df[column_name].apply(lambda x: str(Seq(x).reverse_complement()))\n",
    "    return df\n",
    "    \n",
    "    \n",
    "    \n",
    "# Merging sequences base on the overlping region\n",
    "\n",
    "def merge_sequences(row, min_overlap=12):\n",
    "    \"\"\"Merge forward and reverse reads based on overlapping regions.\"\"\"\n",
    "    f_seq = row['Forward Read']\n",
    "    r_seq = row['Reverse Read']\n",
    "    #Inserting a code in case tthere is no overlap\n",
    "#    no_over_id = \"catcatcatcat\"\n",
    "    max_overlap = min(len(f_seq), len(r_seq))\n",
    "    \n",
    "    for i in range(min_overlap, max_overlap):\n",
    "        if f_seq[-i:] == r_seq[:i]:\n",
    "            return f_seq + r_seq[i:]\n",
    "    return f_seq + r_seq  # Fallback if no overlap is sufficient (Return both reads)\n",
    "\n",
    "#Merge two files\n",
    "def merge_files(file1, file2, output_file):\n",
    "    with open(file1, 'r') as f1, open(file2, 'r') as f2, open(output_file, 'w') as output:\n",
    "        data1 = f1.read()\n",
    "        data2 = f2.read()\n",
    "        output.write(data1)\n",
    "        output.write(data2)\n",
    "\n",
    "        #convert text to fasta again \n",
    "\n",
    "def Convert_TXT_to_Fasta (infile , outfile):\n",
    "    fileInput = open(infile, \"r\")\n",
    "    fileOutput = open(outfile, \"w\")\n",
    "    #Seq counter\n",
    "    count = 1 ;\n",
    "\n",
    "    #Loop through each line in the input file\n",
    "    print (\"Converting\", os.path.basename(infile) ,\" to FASTA...\")\n",
    "    for strLine in fileInput:\n",
    "        if not strLine.isspace() :\n",
    "            #Strip the endline character from each input line\n",
    "            strLine = strLine.rstrip(\"\\n\")\n",
    "    \n",
    "            #Output the header\n",
    "            fileOutput.write(\">\" + str(count) + \"\\n\")\n",
    "            fileOutput.write(strLine + \"\\n\")\n",
    "\n",
    "            count = count + 1\n",
    "    print (\"Done.\")\n",
    "\n",
    "    #Close the input and output file\n",
    "    fileInput.close()\n",
    "    fileOutput.close()\n",
    "\n",
    "    \n",
    "#Filtering out peptides with stop codon from sequensces\n",
    "\n",
    "def Filterstop(Fin , Fout):\n",
    "    print (\"Removing peptides with stop codon in :\", os.path.basename(Fin) ,\" ...\")\n",
    "    fout=open(Fout,'w')\n",
    "\n",
    "    with open(Fin ,'r') as fin:\n",
    "        for line in fin:\n",
    "            line=line.strip()\n",
    "            if '*'not in line:\n",
    "                fout.write(\"%s\\n\" %(line))\n",
    "    fout.close()\n",
    "    print (\"Done.\")\n",
    "\n",
    "    \n",
    "#remove short sequences and NAN values\n",
    "def dataclean (Dataframein):\n",
    "\n",
    "    filtered_df = Dataframein[Dataframein['sequence_length'] >92]\n",
    "\n",
    "    filtered_df['sequence'].replace('', np.nan, inplace=True)\n",
    "    filtered_df.dropna(subset=['sequence'], inplace=True)\n",
    "    filtered_df.dropna(subset=['sequence_length'], inplace=True)\n",
    "    return filtered_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "517476d1-5536-4b87-be97-c92c5650ae2a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Read Number</th>\n",
       "      <th>Forward Read</th>\n",
       "      <th>Reverse Read</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>NTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...</td>\n",
       "      <td>GCCGGTCAAAGAACGAGAATTTGTCGCATCAATTACGTTAGTTGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>NTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...</td>\n",
       "      <td>GCCGGGCAAAGAACGCGAGTTCGTAGCTGGTATCGGCTACGGCTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>TTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...</td>\n",
       "      <td>GCCGGGCAAAGAACGCGAATTTGTTGCCGGCATTGCTTCAGGTAGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>ACGATTTTCTGTATCTCAAGTTGCAAACCGTTCTGAGTGTTTTACC...</td>\n",
       "      <td>ACGATTTTCTGTATCTCAAGTTGCAAACCGTTCTGAGTGTTTTACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>GACGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...</td>\n",
       "      <td>GCCGGGCAAAGAACGCGAGTTTGTCGCCACAATATCTGGCGGGGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313330</th>\n",
       "      <td>313331</td>\n",
       "      <td>ATCAGCGGCGGCGGCCTGGGCACCCAGGTGACCGTGAGCAGCACCCAAG</td>\n",
       "      <td>ATCAGCGGCGGCGGCCTGGGCACCCAGGTGACCGTGAGCAGCACCCAAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313331</th>\n",
       "      <td>313332</td>\n",
       "      <td>ATCAGCGGCGGCGGCCTGGTGCAGGCGGGCGGCAGCCTGCGCCTGA...</td>\n",
       "      <td>CGCCAGGCGCCGGGCAAAGAACGCGAACTGGTAGCCGGCATCACAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313332</th>\n",
       "      <td>313333</td>\n",
       "      <td>TTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...</td>\n",
       "      <td>GTATCGACAGGCGCCGGGCAAAGAACGCGAGTTCGTGGCTGGGATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313333</th>\n",
       "      <td>313334</td>\n",
       "      <td>ATCCAGCTGCAGGAAAGCGGGAAGGAATAAGAAGCAACTAAACGAG...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTATAACGTTGTTTATTTTTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313334</th>\n",
       "      <td>313335</td>\n",
       "      <td>TTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...</td>\n",
       "      <td>GTATCGCCAGGCGCCGGGCAAAGAACGCGAATTAGTTGCGGGAATA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>313335 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Read Number                                       Forward Read  \\\n",
       "0                 1  NTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...   \n",
       "1                 2  NTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...   \n",
       "2                 3  TTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...   \n",
       "3                 4  ACGATTTTCTGTATCTCAAGTTGCAAACCGTTCTGAGTGTTTTACC...   \n",
       "4                 5  GACGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...   \n",
       "...             ...                                                ...   \n",
       "313330       313331  ATCAGCGGCGGCGGCCTGGGCACCCAGGTGACCGTGAGCAGCACCCAAG   \n",
       "313331       313332  ATCAGCGGCGGCGGCCTGGTGCAGGCGGGCGGCAGCCTGCGCCTGA...   \n",
       "313332       313333  TTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...   \n",
       "313333       313334  ATCCAGCTGCAGGAAAGCGGGAAGGAATAAGAAGCAACTAAACGAG...   \n",
       "313334       313335  TTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...   \n",
       "\n",
       "                                             Reverse Read  \n",
       "0       GCCGGTCAAAGAACGAGAATTTGTCGCATCAATTACGTTAGTTGGA...  \n",
       "1       GCCGGGCAAAGAACGCGAGTTCGTAGCTGGTATCGGCTACGGCTCC...  \n",
       "2       GCCGGGCAAAGAACGCGAATTTGTTGCCGGCATTGCTTCAGGTAGC...  \n",
       "3       ACGATTTTCTGTATCTCAAGTTGCAAACCGTTCTGAGTGTTTTACC...  \n",
       "4       GCCGGGCAAAGAACGCGAGTTTGTCGCCACAATATCTGGCGGGGGC...  \n",
       "...                                                   ...  \n",
       "313330  ATCAGCGGCGGCGGCCTGGGCACCCAGGTGACCGTGAGCAGCACCCAAG  \n",
       "313331  CGCCAGGCGCCGGGCAAAGAACGCGAACTGGTAGCCGGCATCACAA...  \n",
       "313332  GTATCGACAGGCGCCGGGCAAAGAACGCGAGTTCGTGGCTGGGATT...  \n",
       "313333  TTTTTTTTTTTTTTTTTTTTTTTTTTATAACGTTGTTTATTTTTTT...  \n",
       "313334  GTATCGCCAGGCGCCGGGCAAAGAACGCGAATTAGTTGCGGGAATA...  \n",
       "\n",
       "[313335 rows x 3 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###Input NGS Files\n",
    "# File paths for revers and forward read\n",
    "forward_fastq_path = \"../NGS-files/ENR234-NbLibNE_S1_L001_R1_001.fastq\"\n",
    "reverse_fastq_path = \"../NGS-files/ENR234-NbLibNE_S1_L001_R2_001.fastq\"\n",
    "\n",
    "# Add reads into a dataFrame\n",
    "read_df = match_reads(forward_fastq_path, reverse_fastq_path)\n",
    "def reverse_complement(df, column_name):\n",
    "    \"\"\"Reverse complement sequences in a specified column of a DataFrame.\"\"\"\n",
    "    df[column_name] = df[column_name].apply(lambda x: str(Seq(x).reverse_complement()))\n",
    "    return df\n",
    "\n",
    "# Display the DataFrame\n",
    "read_df = reverse_complement(read_df, 'Reverse Read')\n",
    "read_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "55d992b3-1314-45f2-bb8f-ccb02bee1500",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Read Number</th>\n",
       "      <th>Forward Read</th>\n",
       "      <th>Reverse Read</th>\n",
       "      <th>Complete Read</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>NTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...</td>\n",
       "      <td>GCCGGTCAAAGAACGAGAATTTGTCGCATCAATTACGTTAGTTGGA...</td>\n",
       "      <td>NTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>NTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...</td>\n",
       "      <td>GCCGGGCAAAGAACGCGAGTTCGTAGCTGGTATCGGCTACGGCTCC...</td>\n",
       "      <td>NTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>TTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...</td>\n",
       "      <td>GCCGGGCAAAGAACGCGAATTTGTTGCCGGCATTGCTTCAGGTAGC...</td>\n",
       "      <td>TTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>ACGATTTTCTGTATCTCAAGTTGCAAACCGTTCTGAGTGTTTTACC...</td>\n",
       "      <td>ACGATTTTCTGTATCTCAAGTTGCAAACCGTTCTGAGTGTTTTACC...</td>\n",
       "      <td>ACGATTTTCTGTATCTCAAGTTGCAAACCGTTCTGAGTGTTTTACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>GACGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...</td>\n",
       "      <td>GCCGGGCAAAGAACGCGAGTTTGTCGCCACAATATCTGGCGGGGGC...</td>\n",
       "      <td>GACGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313330</th>\n",
       "      <td>313331</td>\n",
       "      <td>ATCAGCGGCGGCGGCCTGGGCACCCAGGTGACCGTGAGCAGCACCCAAG</td>\n",
       "      <td>ATCAGCGGCGGCGGCCTGGGCACCCAGGTGACCGTGAGCAGCACCCAAG</td>\n",
       "      <td>ATCAGCGGCGGCGGCCTGGGCACCCAGGTGACCGTGAGCAGCACCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313331</th>\n",
       "      <td>313332</td>\n",
       "      <td>ATCAGCGGCGGCGGCCTGGTGCAGGCGGGCGGCAGCCTGCGCCTGA...</td>\n",
       "      <td>CGCCAGGCGCCGGGCAAAGAACGCGAACTGGTAGCCGGCATCACAA...</td>\n",
       "      <td>ATCAGCGGCGGCGGCCTGGTGCAGGCGGGCGGCAGCCTGCGCCTGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313332</th>\n",
       "      <td>313333</td>\n",
       "      <td>TTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...</td>\n",
       "      <td>GTATCGACAGGCGCCGGGCAAAGAACGCGAGTTCGTGGCTGGGATT...</td>\n",
       "      <td>TTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313333</th>\n",
       "      <td>313334</td>\n",
       "      <td>ATCCAGCTGCAGGAAAGCGGGAAGGAATAAGAAGCAACTAAACGAG...</td>\n",
       "      <td>TTTTTTTTTTTTTTTTTTTTTTTTTTATAACGTTGTTTATTTTTTT...</td>\n",
       "      <td>ATCCAGCTGCAGGAAAGCGGGAAGGAATAAGAAGCAACTAAACGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313334</th>\n",
       "      <td>313335</td>\n",
       "      <td>TTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...</td>\n",
       "      <td>GTATCGCCAGGCGCCGGGCAAAGAACGCGAATTAGTTGCGGGAATA...</td>\n",
       "      <td>TTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>313335 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Read Number                                       Forward Read  \\\n",
       "0                 1  NTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...   \n",
       "1                 2  NTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...   \n",
       "2                 3  TTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...   \n",
       "3                 4  ACGATTTTCTGTATCTCAAGTTGCAAACCGTTCTGAGTGTTTTACC...   \n",
       "4                 5  GACGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...   \n",
       "...             ...                                                ...   \n",
       "313330       313331  ATCAGCGGCGGCGGCCTGGGCACCCAGGTGACCGTGAGCAGCACCCAAG   \n",
       "313331       313332  ATCAGCGGCGGCGGCCTGGTGCAGGCGGGCGGCAGCCTGCGCCTGA...   \n",
       "313332       313333  TTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...   \n",
       "313333       313334  ATCCAGCTGCAGGAAAGCGGGAAGGAATAAGAAGCAACTAAACGAG...   \n",
       "313334       313335  TTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...   \n",
       "\n",
       "                                             Reverse Read  \\\n",
       "0       GCCGGTCAAAGAACGAGAATTTGTCGCATCAATTACGTTAGTTGGA...   \n",
       "1       GCCGGGCAAAGAACGCGAGTTCGTAGCTGGTATCGGCTACGGCTCC...   \n",
       "2       GCCGGGCAAAGAACGCGAATTTGTTGCCGGCATTGCTTCAGGTAGC...   \n",
       "3       ACGATTTTCTGTATCTCAAGTTGCAAACCGTTCTGAGTGTTTTACC...   \n",
       "4       GCCGGGCAAAGAACGCGAGTTTGTCGCCACAATATCTGGCGGGGGC...   \n",
       "...                                                   ...   \n",
       "313330  ATCAGCGGCGGCGGCCTGGGCACCCAGGTGACCGTGAGCAGCACCCAAG   \n",
       "313331  CGCCAGGCGCCGGGCAAAGAACGCGAACTGGTAGCCGGCATCACAA...   \n",
       "313332  GTATCGACAGGCGCCGGGCAAAGAACGCGAGTTCGTGGCTGGGATT...   \n",
       "313333  TTTTTTTTTTTTTTTTTTTTTTTTTTATAACGTTGTTTATTTTTTT...   \n",
       "313334  GTATCGCCAGGCGCCGGGCAAAGAACGCGAATTAGTTGCGGGAATA...   \n",
       "\n",
       "                                            Complete Read  \n",
       "0       NTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...  \n",
       "1       NTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...  \n",
       "2       TTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...  \n",
       "3       ACGATTTTCTGTATCTCAAGTTGCAAACCGTTCTGAGTGTTTTACC...  \n",
       "4       GACGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...  \n",
       "...                                                   ...  \n",
       "313330  ATCAGCGGCGGCGGCCTGGGCACCCAGGTGACCGTGAGCAGCACCC...  \n",
       "313331  ATCAGCGGCGGCGGCCTGGTGCAGGCGGGCGGCAGCCTGCGCCTGA...  \n",
       "313332  TTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...  \n",
       "313333  ATCCAGCTGCAGGAAAGCGGGAAGGAATAAGAAGCAACTAAACGAG...  \n",
       "313334  TTCGTTATTGCTTCAGTTTTAGCACAGGTGCAGCTGCAGGAAAGCG...  \n",
       "\n",
       "[313335 rows x 4 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a new column 'Complete Read' by merging the forward and reverse reads\n",
    "read_df['Complete Read'] = read_df.apply(merge_sequences, axis=1)\n",
    "read_df.to_csv(\"read_df.csv\", index=True)\n",
    "read_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d9a067c3-f577-4d4e-a96b-51f5ffc4df2b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Convert the 'Complete Read's  to a txt file\n",
    "with open(\"complete_reads.txt\", \"w\") as file:\n",
    "    for read in read_df['Complete Read']:\n",
    "        file.write(read + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "baa1dd25-141a-4cd2-8a62-d9b7510277b5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Round 2 ID is  :  GACGTTATTGCT \n",
      " Round 3 ID is :  TTCGTTATTGCT \n",
      " Round 4_V1 ID is :  ATCCAGCTGCAG \n",
      " Round 4_V2 ID is :  ATCAGCGGCGGC\n",
      "Data seperated into R2, R3, R4V1 and R4V2 ! \n"
     ]
    }
   ],
   "source": [
    "#Extracting different rounds of enrichment with specific IDs\n",
    " \n",
    "\n",
    "ENR2_id    = \"GACGTTATTGCT\"\n",
    "ENR3_id    = \"TTCGTTATTGCT\"\n",
    "ENR4_v1_id = \"ATCCAGCTGCAG\"\n",
    "ENR4_v2_id = \"ATCAGCGGCGGC\"\n",
    "nb_id  = \"AGCGGCGGCGGCCTGGTG\"\n",
    "\n",
    "print (\" Round 2 ID is  : \", ENR2_id , \"\\n\" , \"Round 3 ID is : \", ENR3_id , \"\\n\" , \"Round 4_V1 ID is : \", ENR4_v1_id, \"\\n\" , \"Round 4_V2 ID is : \", ENR4_v2_id)\n",
    "                 \n",
    "\n",
    "    \n",
    "def Extract(Data_file , ENR2_id , ENR3_id , ENR4_v1_id ,ENR4_v2_id ):\n",
    "    output_ENR2    = open(\"ENR2.txt\", \"w\")\n",
    "    output_ENR3    = open(\"ENR3.txt\", \"w\")\n",
    "    output_ENR4v1  = open(\"ENR4v1.txt\", \"w\")\n",
    "    output_ENR4v2  = open(\"ENR4v2.txt\", \"w\")\n",
    "    output_Extra  = open(\"Extra.txt\", \"w\")\n",
    "    \n",
    "    with open(Data_file,\"r\") as In_file:\n",
    "        for line in In_file:\n",
    "            seq=line.strip()\n",
    "            \n",
    "            if seq.find(ENR2_id) != -1:\n",
    "                # Code for round2\n",
    "                TagStart = seq.find(nb_id)\n",
    "                sub=seq[TagStart:]\n",
    "#                sub=seq #whole line\n",
    "                output_ENR2.write(\"%s\\n\" %sub)\n",
    "                    \n",
    "            elif seq.find(ENR3_id) != -1:\n",
    "                # Code for round3\n",
    "                TagStart = seq.find(nb_id)\n",
    "                sub=seq[TagStart:]\n",
    "#                sub=seq #whole line\n",
    "                output_ENR3.write(\"%s\\n\" %sub)\n",
    "                    \n",
    "            elif seq.find(ENR4_v1_id) != -1:\n",
    "                # Code for round4-1\n",
    "                TagStart = seq.find(nb_id)\n",
    "                sub=seq[TagStart:]\n",
    "#                sub=seq #whole line\n",
    "                output_ENR4v1.write(\"%s\\n\" %sub)\n",
    "                   \n",
    "            elif seq.find(ENR4_v2_id) != -1:\n",
    "                # Code for round4-2\n",
    "                TagStart = seq.find(nb_id)\n",
    "                sub=seq[TagStart:]\n",
    "#               sub=seq #whole line\n",
    "                output_ENR4v2.write(\"%s\\n\" %sub)\n",
    "                   \n",
    "            else:\n",
    "                # Default case, Put them in extra\n",
    "                sub=seq  #whole line   \n",
    "                output_Extra.write(\"%s\\n\" %sub)\n",
    "        \n",
    "    output_ENR2.close()\n",
    "    output_ENR3.close()\n",
    "    output_ENR4v1.close()\n",
    "    output_ENR4v2.close()\n",
    "    output_Extra.close()\n",
    "\n",
    "    print (\"Data seperated into R2, R3, R4V1 and R4V2 ! \")\n",
    "\n",
    "        \n",
    "Extract(\"complete_reads.txt\" , ENR2_id , ENR3_id , ENR4_v1_id ,ENR4_v2_id ) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a7206958-548f-4756-adbf-f4be75fe06ba",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Data lines: 313335\n",
      "Number of Total lines: 313335\n",
      "Extraction is acurate, All lines are included in 5 files : Round2 : 88968 , Round 3 : 116160, Round 4 part 1:43992, Round 4 part 2:41082 and extra:23133 \n",
      "(NBs) : 92.61716692996313 % and Extra: 7.382833070036862\n"
     ]
    }
   ],
   "source": [
    "# counting lines and checking acuracy:\n",
    "\n",
    "def count_lines(file_path):\n",
    "    with open(file_path, 'r') as file:\n",
    "        line_count = sum(1 for _ in file)\n",
    "    return line_count\n",
    "\n",
    "#Usage:\n",
    "Data_lines = count_lines(\"complete_reads.txt\")\n",
    "ENR2_lines = count_lines(\"ENR2.txt\")\n",
    "ENR3_lines = count_lines(\"ENR3.txt\")\n",
    "ENR4v1_lines = count_lines(\"ENR4v1.txt\")\n",
    "ENR4v2_lines = count_lines(\"ENR4v2.txt\")\n",
    "Extra_lines = count_lines(\"Extra.txt\")\n",
    "\n",
    "\n",
    "Total = ENR2_lines + ENR3_lines + ENR4v1_lines + ENR4v2_lines + Extra_lines\n",
    "\n",
    "print(f\"Number of Data lines: {Data_lines}\")\n",
    "print(f\"Number of Total lines: {Total}\")\n",
    "\n",
    "if Data_lines == Total :\n",
    "    print(f\"Extraction is acurate, All lines are included in 5 files : Round2 : {ENR2_lines} , Round 3 : {ENR3_lines}, Round 4 part 1:{ENR4v1_lines}, Round 4 part 2:{ENR4v2_lines} and extra:{Extra_lines} \")\n",
    "else :\n",
    "    print(f\"Some lines are missing\")\n",
    "\n",
    "print(f\"(NBs) : {((ENR2_lines + ENR3_lines + ENR4v1_lines + ENR4v2_lines)/Total)* 100} % and Extra: {((Extra_lines)/Total)* 100}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8ddcbd9e-ac88-424a-8129-cbf096c10bfb",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENR4v1 and ENR4v2 Merged susccefully : ENR4.txt\n"
     ]
    }
   ],
   "source": [
    "#Merge round 4 two files (we submited 2 NGS samples for round to increas the coverage)     \n",
    "merge_files('ENR4v1.txt', 'ENR4v2.txt', 'ENR4.txt')\n",
    "print(\"ENR4v1 and ENR4v2 Merged susccefully : ENR4.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5c651932-a1f5-42e6-9c21-b092486ed77b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converting ENR2.txt  to FASTA...\n",
      "Done.\n",
      "Converting ENR3.txt  to FASTA...\n",
      "Done.\n",
      "Converting ENR4.txt  to FASTA...\n",
      "Done.\n",
      "Converting Extra.txt  to FASTA...\n",
      "Done.\n"
     ]
    }
   ],
   "source": [
    "# Convert txt to Fasta\n",
    "Convert_TXT_to_Fasta(\"ENR2.txt\",\"ENR2.fasta\")\n",
    "Convert_TXT_to_Fasta(\"ENR3.txt\",\"ENR3.fasta\")\n",
    "Convert_TXT_to_Fasta(\"ENR4.txt\",\"ENR4.fasta\")\n",
    "Convert_TXT_to_Fasta(\"Extra.txt\",\"Extra.fasta\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8246bfdc-6207-4958-8479-56594b3c3226",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Translating R2 from DNA to Protein...\n",
      "Done.\n",
      "Translating R3 from DNA to Protein...\n",
      "Done.\n",
      "Translating R4 from DNA to Protein...\n",
      "Done.\n",
      "Translating extra from DNA to Protein...\n",
      "Done.\n"
     ]
    }
   ],
   "source": [
    "#convert DNA Alphabet to Protein, translate to amino acids\n",
    "\n",
    "#Round2\n",
    "print (\"Translating R2 from DNA to Protein...\")\n",
    "def translation(nuc):\n",
    "    return SeqRecord(seq=nuc.seq.translate(), id = nuc.id)\n",
    "\n",
    "proteins = (translation(nuc_rec) for nuc_rec in SeqIO.parse(\"ENR2.fasta\",'fasta'))\n",
    "SeqIO.write(proteins, 'protENR2.fasta','fasta')\n",
    "print (\"Done.\")\n",
    "\n",
    "#Round3\n",
    "print (\"Translating R3 from DNA to Protein...\")\n",
    "def translation(nuc):\n",
    "    return SeqRecord(seq=nuc.seq.translate(), id = nuc.id)\n",
    "\n",
    "proteins = (translation(nuc_rec) for nuc_rec in SeqIO.parse(\"ENR3.fasta\",'fasta'))\n",
    "SeqIO.write(proteins, 'protENR3.fasta','fasta')\n",
    "print (\"Done.\")\n",
    "\n",
    "#Round 4\n",
    "print (\"Translating R4 from DNA to Protein...\")\n",
    "def translation(nuc):\n",
    "    return SeqRecord(seq=nuc.seq.translate(), id = nuc.id)\n",
    "\n",
    "proteins = (translation(nuc_rec) for nuc_rec in SeqIO.parse(\"ENR4.fasta\",'fasta'))\n",
    "SeqIO.write(proteins, 'protENR4.fasta','fasta')\n",
    "print (\"Done.\")\n",
    "\n",
    "#EXTRA Seq\n",
    "print (\"Translating extra from DNA to Protein...\")\n",
    "def translation(nuc):\n",
    "    return SeqRecord(seq=nuc.seq.translate(), id = nuc.id)\n",
    "\n",
    "proteins = (translation(nuc_rec) for nuc_rec in SeqIO.parse(\"Extra.fasta\",'fasta'))\n",
    "SeqIO.write(proteins, 'protExtra.fasta','fasta')\n",
    "print (\"Done.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "bd23a339-9252-4300-88cb-e5403464207b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removing peptides with stop codon in : protENR2.fasta  ...\n",
      "Done.\n",
      "Removing peptides with stop codon in : protENR3.fasta  ...\n",
      "Done.\n",
      "Removing peptides with stop codon in : protENR4.fasta  ...\n",
      "Done.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence_length</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>113</td>\n",
       "      <td>SGGGLVQAGGSLRLSCAASGTIFGGVDMGWYRQAPGKEREFVAAIA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>121</td>\n",
       "      <td>SGGGLVQAGGSLRLSCAASGNISYFGGMGWYRQAPGKERELVAGIA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85069</th>\n",
       "      <td>85070</td>\n",
       "      <td>118</td>\n",
       "      <td>SGGGLVQAGGSLRLSCAASGTISYGWHMGWYRQAPGKERELVATIN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85070</th>\n",
       "      <td>85071</td>\n",
       "      <td>60</td>\n",
       "      <td>SGGGLVQAGGSLRLSCAASGSISDDDAPGKYSLDDSVYIYWYWGQG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85071</th>\n",
       "      <td>85072</td>\n",
       "      <td>114</td>\n",
       "      <td>SGGGLVQAGGSLRLSCAASGTIFHGQFMGWYRQAPGKERELVATIG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85072</th>\n",
       "      <td>85073</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85073</th>\n",
       "      <td>85074</td>\n",
       "      <td>114</td>\n",
       "      <td>SGGGLVQAGGSLRLSCAASGYIFGGGPMGWYRQAPGKERELVAGIT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>85074 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          id  sequence_length  \\\n",
       "0          1                0   \n",
       "1          2              113   \n",
       "2          3                0   \n",
       "3          4                0   \n",
       "4          5              121   \n",
       "...      ...              ...   \n",
       "85069  85070              118   \n",
       "85070  85071               60   \n",
       "85071  85072              114   \n",
       "85072  85073                0   \n",
       "85073  85074              114   \n",
       "\n",
       "                                                sequence  \n",
       "0                                                         \n",
       "1      SGGGLVQAGGSLRLSCAASGTIFGGVDMGWYRQAPGKEREFVAAIA...  \n",
       "2                                                         \n",
       "3                                                         \n",
       "4      SGGGLVQAGGSLRLSCAASGNISYFGGMGWYRQAPGKERELVAGIA...  \n",
       "...                                                  ...  \n",
       "85069  SGGGLVQAGGSLRLSCAASGTISYGWHMGWYRQAPGKERELVATIN...  \n",
       "85070  SGGGLVQAGGSLRLSCAASGSISDDDAPGKYSLDDSVYIYWYWGQG...  \n",
       "85071  SGGGLVQAGGSLRLSCAASGTIFHGQFMGWYRQAPGKERELVATIG...  \n",
       "85072                                                     \n",
       "85073  SGGGLVQAGGSLRLSCAASGYIFGGGPMGWYRQAPGKERELVAGIT...  \n",
       "\n",
       "[85074 rows x 3 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#filtering out peptides with stop codon from sequensces\n",
    "\n",
    "Filterstop(\"protENR2.fasta\" , \"protENR2f.fasta\")\n",
    "Filterstop(\"protENR3.fasta\" , \"protENR3f.fasta\")\n",
    "Filterstop(\"protENR4.fasta\" , \"protENR4f.fasta\")\n",
    "\n",
    "#show and store All data in data frames :\n",
    "dataR2   = read_fasta(\"protENR2f.fasta\", columns=[\"id\",\"sequence_length\", \"sequence\"])\n",
    "dataR3   = read_fasta(\"protENR3f.fasta\", columns=[\"id\",\"sequence_length\", \"sequence\"])\n",
    "dataR4   = read_fasta(\"protENR4f.fasta\", columns=[\"id\",\"sequence_length\", \"sequence\"])\n",
    "dataR4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3fa3fa0a-eb1a-4d2c-8a0b-a113e8892fa6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence_length</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>113</td>\n",
       "      <td>SGGGLVQAGGSLRLSCAASGTIFGGVDMGWYRQAPGKEREFVAAIA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>121</td>\n",
       "      <td>SGGGLVQAGGSLRLSCAASGNISYFGGMGWYRQAPGKERELVAGIA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16</td>\n",
       "      <td>117</td>\n",
       "      <td>SGGGLVQAGGSLRLSCAASGNIFRVREMGWYRQAPGKEREFVAGIA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>23</td>\n",
       "      <td>120</td>\n",
       "      <td>SGGGLVQAGGSLRLSCAASGTISPYRWMGWYRQAPGKEREFVAGIG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>32</td>\n",
       "      <td>121</td>\n",
       "      <td>SGGGLVQAGGSLRLSCAASGYIFRKQRMGWYRQAPGKERELVASIS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85065</th>\n",
       "      <td>85066</td>\n",
       "      <td>118</td>\n",
       "      <td>SGGGLVQAGGSLRLSCAASGTISYGWHMGWYRQAPGKEREFVASID...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85066</th>\n",
       "      <td>85067</td>\n",
       "      <td>117</td>\n",
       "      <td>SGGGLVQAGGSLRLSCAASGYIFGGGPMGWYRQAPGKEREFVATID...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85069</th>\n",
       "      <td>85070</td>\n",
       "      <td>118</td>\n",
       "      <td>SGGGLVQAGGSLRLSCAASGTISYGWHMGWYRQAPGKERELVATIN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85071</th>\n",
       "      <td>85072</td>\n",
       "      <td>114</td>\n",
       "      <td>SGGGLVQAGGSLRLSCAASGTIFHGQFMGWYRQAPGKERELVATIG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85073</th>\n",
       "      <td>85074</td>\n",
       "      <td>114</td>\n",
       "      <td>SGGGLVQAGGSLRLSCAASGYIFGGGPMGWYRQAPGKERELVAGIT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36235 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          id  sequence_length  \\\n",
       "1          2              113   \n",
       "4          5              121   \n",
       "15        16              117   \n",
       "22        23              120   \n",
       "31        32              121   \n",
       "...      ...              ...   \n",
       "85065  85066              118   \n",
       "85066  85067              117   \n",
       "85069  85070              118   \n",
       "85071  85072              114   \n",
       "85073  85074              114   \n",
       "\n",
       "                                                sequence  \n",
       "1      SGGGLVQAGGSLRLSCAASGTIFGGVDMGWYRQAPGKEREFVAAIA...  \n",
       "4      SGGGLVQAGGSLRLSCAASGNISYFGGMGWYRQAPGKERELVAGIA...  \n",
       "15     SGGGLVQAGGSLRLSCAASGNIFRVREMGWYRQAPGKEREFVAGIA...  \n",
       "22     SGGGLVQAGGSLRLSCAASGTISPYRWMGWYRQAPGKEREFVAGIG...  \n",
       "31     SGGGLVQAGGSLRLSCAASGYIFRKQRMGWYRQAPGKERELVASIS...  \n",
       "...                                                  ...  \n",
       "85065  SGGGLVQAGGSLRLSCAASGTISYGWHMGWYRQAPGKEREFVASID...  \n",
       "85066  SGGGLVQAGGSLRLSCAASGYIFGGGPMGWYRQAPGKEREFVATID...  \n",
       "85069  SGGGLVQAGGSLRLSCAASGTISYGWHMGWYRQAPGKERELVATIN...  \n",
       "85071  SGGGLVQAGGSLRLSCAASGTIFHGQFMGWYRQAPGKERELVATIG...  \n",
       "85073  SGGGLVQAGGSLRLSCAASGYIFGGGPMGWYRQAPGKERELVAGIT...  \n",
       "\n",
       "[36235 rows x 3 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#remove short sequences and NAN values\n",
    "\n",
    "dataR2_C= dataclean(dataR2)\n",
    "dataR3_C= dataclean(dataR3)\n",
    "dataR4_C= dataclean(dataR4)\n",
    "\n",
    "dataR2= dataR2_C\n",
    "dataR3= dataR3_C\n",
    "dataR4= dataR4_C\n",
    "\n",
    "# Deleting temp dataframes to make free storage\n",
    "del dataR2_C\n",
    "del dataR3_C\n",
    "del dataR4_C\n",
    "\n",
    "dataR4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "12490826-2db5-429c-9131-069698c6a60d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "dftempR2  = dataR2.loc[:, ['sequence_length' , 'sequence']]\n",
    "dftempR3  = dataR3.loc[:, ['sequence_length' , 'sequence']]\n",
    "dftempR4 = dataR4.loc[:, ['sequence_length' , 'sequence']]\n",
    "\n",
    "df_R2 = dftempR2\n",
    "df_R3 = dftempR3\n",
    "df_R4 = dftempR4\n",
    "\n",
    "del dftempR2\n",
    "del dftempR3\n",
    "del dftempR4\n",
    "\n",
    "\n",
    "## Finding Sequensce of Nanobodies from the reads and fix their starting sequensces\n",
    "\n",
    "df_R2['sequence'] = df_R2['sequence'].apply(lambda x: x.split('VTVSS')[0] + 'VTVSS' if 'VTVSS' in x else x)\n",
    "df_R3['sequence'] = df_R3['sequence'].apply(lambda x: x.split('VTVSS')[0] + 'VTVSS' if 'VTVSS' in x else x)\n",
    "df_R4['sequence'] = df_R4['sequence'].apply(lambda x: x.split('VTVSS')[0] + 'VTVSS' if 'VTVSS' in x else x)\n",
    "\n",
    "df_R2['sequence_length'] = df_R2['sequence'].apply(lambda x: len(x) if isinstance(x, str) else 0)\n",
    "df_R3['sequence_length'] = df_R3['sequence'].apply(lambda x: len(x) if isinstance(x, str) else 0)\n",
    "df_R4['sequence_length'] = df_R4['sequence'].apply(lambda x: len(x) if isinstance(x, str) else 0)\n",
    "\n",
    "# remove very long or very short reads to improve acuracy \n",
    "smin = 92\n",
    "smax= 125\n",
    "\n",
    "df_R2 = df_R2[df_R2[\"sequence_length\"] >= smin]\n",
    "df_R2 = df_R2[df_R2[\"sequence_length\"] <= smax]\n",
    "\n",
    "df_R3 = df_R3[df_R3[\"sequence_length\"] >= smin]\n",
    "df_R3 = df_R3[df_R3[\"sequence_length\"] <= smax]\n",
    "\n",
    "df_R4 = df_R4[df_R4[\"sequence_length\"] >= smin]\n",
    "df_R4 = df_R4[df_R4[\"sequence_length\"] <= smax]\n",
    "\n",
    "#make the original sequence format of NB (Add amino acids in front od reads)\n",
    "df_R2.loc[:, 'sequence'] = df_R2['sequence'].apply(lambda x: \"QVQLQE\" + x)\n",
    "df_R3.loc[:, 'sequence'] = df_R3['sequence'].apply(lambda x: \"QVQLQE\" + x)\n",
    "df_R4.loc[:, 'sequence'] = df_R4['sequence'].apply(lambda x: \"QVQLQE\" + x)\n",
    "\n",
    "df_R2['sequence_length'] = df_R2['sequence'].apply(lambda x: len(x) if isinstance(x, str) else 0)\n",
    "df_R3['sequence_length'] = df_R3['sequence'].apply(lambda x: len(x) if isinstance(x, str) else 0)\n",
    "df_R4['sequence_length'] = df_R4['sequence'].apply(lambda x: len(x) if isinstance(x, str) else 0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c2403a8a-b620-4da3-b8e6-f0651210434f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|# # # # # # # # # # # # # # # # # #|\n",
      "|Round 2 Unique Sequences :  26604  |\n",
      "|-----------------------------------|\n",
      "|Round 3 Unique Sequences :  22653  |\n",
      "|-----------------------------------|\n",
      "|Round 4 Unique Sequences :  16006  |\n",
      "|# # # # # # # # # # # # # # # # # #|\n"
     ]
    }
   ],
   "source": [
    "#grouping in order to count each sequenc repeats the data\n",
    "\n",
    "dftempR2t = df_R2\n",
    "dftempR3t = df_R3\n",
    "dftempR4t = df_R4\n",
    "\n",
    "del df_R2\n",
    "del df_R3\n",
    "del df_R4\n",
    "\n",
    "df_R2 = dftempR2t.groupby(dftempR2t.columns.tolist(), as_index=False).size()\n",
    "df_R3 = dftempR3t.groupby(dftempR3t.columns.tolist(), as_index=False).size()\n",
    "df_R4 = dftempR4t.groupby(dftempR4t.columns.tolist(), as_index=False).size()\n",
    "\n",
    "\n",
    "#deleting temp dataframes to free ram space\n",
    "del dftempR2t\n",
    "del dftempR3t\n",
    "del dftempR4t\n",
    "\n",
    "df_R2.sort_values(by= \"size\", ascending= False, inplace=True)\n",
    "\n",
    "df_R3.sort_values(by= \"size\", ascending= False, inplace=True)\n",
    "\n",
    "df_R4.sort_values(by= \"size\", ascending= False, inplace=True)\n",
    "\n",
    "print (\"|# # # # # # # # # # # # # # # # # #|\")\n",
    "print (\"|Round 2 Unique Sequences : \", df_R2.shape[0],\" |\")\n",
    "print (\"|-----------------------------------|\")\n",
    "print (\"|Round 3 Unique Sequences : \", df_R3.shape[0],\" |\")\n",
    "print (\"|-----------------------------------|\")\n",
    "print (\"|Round 4 Unique Sequences : \", df_R4.shape[0],\" |\")\n",
    "print (\"|# # # # # # # # # # # # # # # # # #|\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "310184e9-6c7e-4bcc-841d-f2fa30e7c206",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NGS Enrichment Analysis done sucsesfully.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>size_R2</th>\n",
       "      <th>size_R3</th>\n",
       "      <th>size_R4</th>\n",
       "      <th>Enrichment R3/R2</th>\n",
       "      <th>Enrichment R4/R3</th>\n",
       "      <th>Enrichment R4/R2</th>\n",
       "      <th>sequence_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>QVQLQESGGGLVQAGGSLRLSCAASGSIFTDTYMGWYRQAPGKERE...</td>\n",
       "      <td>117.0</td>\n",
       "      <td>792.0</td>\n",
       "      <td>872.0</td>\n",
       "      <td>6.769231</td>\n",
       "      <td>1.101010</td>\n",
       "      <td>7.452991</td>\n",
       "      <td>122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>QVQLQESGGGLVQAGGSLRLSCAASGYIFGGGPMGWYRQAPGKERE...</td>\n",
       "      <td>281.0</td>\n",
       "      <td>1310.0</td>\n",
       "      <td>793.0</td>\n",
       "      <td>4.661922</td>\n",
       "      <td>0.605344</td>\n",
       "      <td>2.822064</td>\n",
       "      <td>126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>QVQLQESGGGLVQAGGSLRLSCAASGTIFHGQFMGWYRQAPGKERE...</td>\n",
       "      <td>312.0</td>\n",
       "      <td>3453.0</td>\n",
       "      <td>682.0</td>\n",
       "      <td>11.067308</td>\n",
       "      <td>0.197509</td>\n",
       "      <td>2.185897</td>\n",
       "      <td>118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>QVQLQESGGGLVQAGDSLRLSCAASGSIFSYYAMGWYRQAPGKERE...</td>\n",
       "      <td>527.0</td>\n",
       "      <td>1838.0</td>\n",
       "      <td>655.0</td>\n",
       "      <td>3.487666</td>\n",
       "      <td>0.356366</td>\n",
       "      <td>1.242884</td>\n",
       "      <td>118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>QVQLQESGGGLVQAGGSLRLSCAASGYIFVGTYMGWYRQAPGKERE...</td>\n",
       "      <td>69.0</td>\n",
       "      <td>573.0</td>\n",
       "      <td>638.0</td>\n",
       "      <td>8.304348</td>\n",
       "      <td>1.113438</td>\n",
       "      <td>9.246377</td>\n",
       "      <td>122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30622</th>\n",
       "      <td>QVQLQESGGGLVQAGGSLRLSCAASGNIFRPGTMGWYRQAPGKERE...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30623</th>\n",
       "      <td>QVQLQESGGGLVQAGGSLRLSCAASGTISRLPSMGWYRQAPGKERE...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14789</th>\n",
       "      <td>QVQLQESGGGLVQAGGSLRLSCAASGSISTGDGMGWYRQAPGKERE...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30625</th>\n",
       "      <td>QVQLQESGGGLVQAGGSLRLSCAASGYISSSQSMGWYRQAPGKERE...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30240</th>\n",
       "      <td>QVQLQESGGGLVQAGGSLRLSCAASGTISTAGRMGWYRQAPGKERE...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>122</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60480 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                sequence  size_R2  size_R3  \\\n",
       "40     QVQLQESGGGLVQAGGSLRLSCAASGSIFTDTYMGWYRQAPGKERE...    117.0    792.0   \n",
       "7      QVQLQESGGGLVQAGGSLRLSCAASGYIFGGGPMGWYRQAPGKERE...    281.0   1310.0   \n",
       "5      QVQLQESGGGLVQAGGSLRLSCAASGTIFHGQFMGWYRQAPGKERE...    312.0   3453.0   \n",
       "1      QVQLQESGGGLVQAGDSLRLSCAASGSIFSYYAMGWYRQAPGKERE...    527.0   1838.0   \n",
       "88     QVQLQESGGGLVQAGGSLRLSCAASGYIFVGTYMGWYRQAPGKERE...     69.0    573.0   \n",
       "...                                                  ...      ...      ...   \n",
       "30622  QVQLQESGGGLVQAGGSLRLSCAASGNIFRPGTMGWYRQAPGKERE...      1.0      2.0   \n",
       "30623  QVQLQESGGGLVQAGGSLRLSCAASGTISRLPSMGWYRQAPGKERE...      1.0      2.0   \n",
       "14789  QVQLQESGGGLVQAGGSLRLSCAASGSISTGDGMGWYRQAPGKERE...      1.0      0.0   \n",
       "30625  QVQLQESGGGLVQAGGSLRLSCAASGYISSSQSMGWYRQAPGKERE...      1.0      2.0   \n",
       "30240  QVQLQESGGGLVQAGGSLRLSCAASGTISTAGRMGWYRQAPGKERE...      1.0      2.0   \n",
       "\n",
       "       size_R4  Enrichment R3/R2  Enrichment R4/R3  Enrichment R4/R2  \\\n",
       "40       872.0          6.769231          1.101010          7.452991   \n",
       "7        793.0          4.661922          0.605344          2.822064   \n",
       "5        682.0         11.067308          0.197509          2.185897   \n",
       "1        655.0          3.487666          0.356366          1.242884   \n",
       "88       638.0          8.304348          1.113438          9.246377   \n",
       "...        ...               ...               ...               ...   \n",
       "30622      0.0          2.000000          0.000000          0.000000   \n",
       "30623      0.0          2.000000          0.000000          0.000000   \n",
       "14789      0.0          0.000000               NaN          0.000000   \n",
       "30625      0.0          2.000000          0.000000          0.000000   \n",
       "30240      0.0          2.000000          0.000000          0.000000   \n",
       "\n",
       "       sequence_length  \n",
       "40                 122  \n",
       "7                  126  \n",
       "5                  118  \n",
       "1                  118  \n",
       "88                 122  \n",
       "...                ...  \n",
       "30622              118  \n",
       "30623              122  \n",
       "14789              124  \n",
       "30625              126  \n",
       "30240              122  \n",
       "\n",
       "[60480 rows x 8 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Enrichment\n",
    "\n",
    "def Enrichment(df_naive, df_mid, df_sorted, outfile):\n",
    "    # Drop unused column and prepare dataframes\n",
    "    df1 = df_naive.drop(['sequence_length'], axis=1)\n",
    "    df2 = df_mid.drop(['sequence_length'], axis=1)\n",
    "    df3 = df_sorted.drop(['sequence_length'], axis=1)\n",
    "    \n",
    "    # Merge the first two dataframes\n",
    "    merged_df1 = pd.merge(df1, df2, on='sequence', how='outer', suffixes=('_R2', '_R3'))\n",
    "    merged_df1.fillna({'size_R2': 1, 'size_R3': 0}, inplace=True)\n",
    "    \n",
    "    # Merge with the third dataframe\n",
    "    merged_df = pd.merge(merged_df1, df3, on='sequence', how='outer')\n",
    "    merged_df.rename(columns={'size': 'size_R4'}, inplace=True)\n",
    "    merged_df.fillna({'size_R4': 0 , 'size_R2':1 , 'size_R3':1}, inplace=True)\n",
    "    \n",
    "    # Compute enrichment ratios and log values\n",
    "    merged_df[\"Enrichment R3/R2\"] = merged_df[\"size_R3\"] / merged_df[\"size_R2\"]\n",
    "    merged_df[\"Enrichment R4/R3\"] = merged_df[\"size_R4\"] / merged_df[\"size_R3\"]\n",
    "    merged_df[\"Enrichment R4/R2\"] = merged_df[\"size_R4\"] / merged_df[\"size_R2\"]    \n",
    "    # Save to file and return\n",
    "    merged_df.sort_values(by=\"size_R4\", ascending=False, inplace=True)\n",
    "    \n",
    "    merged_df['sequence_length'] = merged_df['sequence'].apply(lambda x: len(x) if isinstance(x, str) else 0)\n",
    "                                                \n",
    "    outdf = merged_df\n",
    "#    outdf.to_csv(outfile, index=True) if you want to save the dataframe here uncomment\n",
    "    print(\"NGS Enrichment Analysis done sucsesfully.\")\n",
    "\n",
    "    return outdf\n",
    "    \n",
    "    \n",
    "NBEnrichment = Enrichment (df_R2, df_R3,df_R4,'HARP_NGS_NB_Enrichment.csv')\n",
    "NBEnrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "b51467cc-016a-4d4a-8d38-4532c9581a57",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Spliting CDRS\n",
    "\n",
    "#CDR 1\n",
    "df_R2['CDR1'] = df_R2['sequence'].apply(lambda x: x.split('ASG')[1].split('MGWY')[0] if ('ASG' in x and 'MGWY' in x) else 1)\n",
    "#CDR 2\n",
    "df_R2['CDR2'] = df_R2['sequence'].apply(lambda x: x.split('ERE')[1].split('YADS')[0] if ('ERE' in x and 'YADS' in x) else 1)\n",
    "#CDR3\n",
    "df_R2['CDR3'] = df_R2['sequence'].apply(lambda x: x.split('YCA')[1].split('YWGQ')[0] if ('YCA' in x and 'YWGQ' in x) else 1)\n",
    "\n",
    "\n",
    "#CDR 1\n",
    "df_R3['CDR1'] = df_R3['sequence'].apply(lambda x: x.split('ASG')[1].split('MGWY')[0] if ('ASG' in x and 'MGWY' in x) else 1)\n",
    "#CDR 2\n",
    "df_R3['CDR2'] = df_R3['sequence'].apply(lambda x: x.split('ERE')[1].split('YADS')[0] if ('ERE' in x and 'YADS' in x) else 1)\n",
    "#CDR3\n",
    "df_R3['CDR3'] = df_R3['sequence'].apply(lambda x: x.split('YCA')[1].split('YWGQ')[0] if ('YCA' in x and 'YWGQ' in x) else 1)\n",
    "\n",
    "\n",
    "#CDR 1\n",
    "df_R4['CDR1'] = df_R4['sequence'].apply(lambda x: x.split('ASG')[1].split('MGWY')[0] if ('ASG' in x and 'MGWY' in x) else 1)\n",
    "#CDR 2\n",
    "df_R4['CDR2'] = df_R4['sequence'].apply(lambda x: x.split('ERE')[1].split('YADS')[0] if ('ERE' in x and 'YADS' in x) else 1)\n",
    "#CDR3\n",
    "df_R4['CDR3'] = df_R4['sequence'].apply(lambda x: x.split('YCA')[1].split('YWGQ')[0] if ('YCA' in x and 'YWGQ' in x) else 1)\n",
    "\n",
    "#CDR 1\n",
    "NBEnrichment['CDR1'] = NBEnrichment['sequence'].apply(lambda x: x.split('ASG')[1].split('MGWY')[0] if ('ASG' in x and 'MGWY' in x) else 1)\n",
    "#CDR 2\n",
    "NBEnrichment['CDR2'] = NBEnrichment['sequence'].apply(lambda x: x.split('ERE')[1].split('YADS')[0] if ('ERE' in x and 'YADS' in x) else 1)\n",
    "#CDR3\n",
    "NBEnrichment['CDR3'] = NBEnrichment['sequence'].apply(lambda x: x.split('YCA')[1].split('YWGQ')[0] if ('YCA' in x and 'YWGQ' in x) else 1)\n",
    "\n",
    "\n",
    "## handling 1 in data:\n",
    "##R1\n",
    "# First step: Modify 'CDR1' for rows where 'CDR1' is 1, try finding the sequence between 'SG' and 'MG'\n",
    "df_R2.loc[:, 'CDR1'] = df_R2.apply(lambda row: row['sequence'].split('SG')[1].split('MG')[0] if row['CDR1'] == 1 and 'SG' in row['sequence'] and 'MG' in row['sequence'] else row['CDR1'], axis=1)\n",
    "\n",
    "# Second step: Modify 'CDR2' for rows where 'CDR2' is 1, try finding the sequence between 'RE' and 'YA'\n",
    "df_R2.loc[:, 'CDR2'] = df_R2.apply(lambda row: row['sequence'].split('RE')[1].split('YA')[0] if row['CDR2'] == 1 and 'RE' in row['sequence'] and 'YA' in row['sequence'] else row['CDR2'], axis=1)\n",
    "\n",
    "df_R2.loc[:, 'CDR3'] = df_R2.apply(lambda row: row['sequence'].split('YCA')[1].split('YWG')[0] if row['CDR3'] == 1 and 'YCA' in row['sequence'] and 'YWG' in row['sequence'] else row['CDR3'], axis=1)\n",
    "\n",
    "# Third step: Modify 'CDR3' for rows where 'CDR3' is 1, try finding the sequence between 'CA' and 'YW'\n",
    "df_R2.loc[:, 'CDR3'] = df_R2.apply(lambda row: row['sequence'].split('YCA')[1].split('YW')[0] if row['CDR3'] == 1 and 'YCA' in row['sequence'] and 'YW' in row['sequence'] else row['CDR3'], axis=1)\n",
    "\n",
    "\n",
    "# Fourth step: Modify 'CDR1' for rows where 'CDR1' is still 1, extract sequence positions 19 to 27\n",
    "df_R2.loc[:, 'CDR1'] = df_R2.apply(lambda row: row['sequence'][19:27] if row['CDR1'] == 1 else row['CDR1'], axis=1)\n",
    "\n",
    "# Fifth step: Modify 'CDR2' for rows where 'CDR2' is still 1, extract sequence positions 39 to 52\n",
    "df_R2.loc[:, 'CDR2'] = df_R2.apply(lambda row: row['sequence'][39:52] if row['CDR2'] == 1 else row['CDR2'], axis=1)\n",
    "\n",
    "##R3\n",
    "# Modify 'CDR1' for rows where 'CDR1' is 1, try finding the sequence between 'SG' and 'MG'\n",
    "df_R3.loc[:, 'CDR1'] = df_R3.apply(lambda row: row['sequence'].split('SG')[1].split('MG')[0] if row['CDR1'] == 1 and 'SG' in row['sequence'] and 'MG' in row['sequence'] else row['CDR1'], axis=1)\n",
    "\n",
    "# Second step: Modify 'CDR2' for rows where 'CDR2' is 1, try finding the sequence between 'RE' and 'YA'\n",
    "df_R3.loc[:, 'CDR2'] = df_R3.apply(lambda row: row['sequence'].split('RE')[1].split('YA')[0] if row['CDR2'] == 1 and 'RE' in row['sequence'] and 'YA' in row['sequence'] else row['CDR2'], axis=1)\n",
    "\n",
    "df_R3.loc[:, 'CDR3'] = df_R3.apply(lambda row: row['sequence'].split('YCA')[1].split('YWG')[0] if row['CDR3'] == 1 and 'YCA' in row['sequence'] and 'YWG' in row['sequence'] else row['CDR3'], axis=1)\n",
    "# Third step: Modify 'CDR3' for rows where 'CDR3' is 1, try finding the sequence between 'CA' and 'YW'\n",
    "df_R3.loc[:, 'CDR3'] = df_R3.apply(lambda row: row['sequence'].split('YCA')[1].split('YW')[0] if row['CDR3'] == 1 and 'YCA' in row['sequence'] and 'YW' in row['sequence'] else row['CDR3'], axis=1)\n",
    "\n",
    "# Fourth step: Modify 'CDR1' for rows where 'CDR1' is still 1, extract sequence positions 19 to 27\n",
    "df_R3.loc[:, 'CDR1'] = df_R3.apply(lambda row: row['sequence'][19:27] if row['CDR1'] == 1 else row['CDR1'], axis=1)\n",
    "\n",
    "# Fifth step: Modify 'CDR2' for rows where 'CDR2' is still 1, extract sequence positions 39 to 52\n",
    "df_R3.loc[:, 'CDR2'] = df_R3.apply(lambda row: row['sequence'][39:52] if row['CDR2'] == 1 else row['CDR2'], axis=1)\n",
    "\n",
    "##R4\n",
    "# First step: Modify 'CDR1' for rows where 'CDR1' is 1, try finding the sequence between 'SG' and 'MG'\n",
    "df_R4.loc[:, 'CDR1'] = df_R4.apply(lambda row: row['sequence'].split('SG')[1].split('MG')[0] if row['CDR1'] == 1 and 'SG' in row['sequence'] and 'MG' in row['sequence'] else row['CDR1'], axis=1)\n",
    "\n",
    "# Second step: Modify 'CDR2' for rows where 'CDR2' is 1, try finding the sequence between 'RE' and 'YA'\n",
    "df_R4.loc[:, 'CDR2'] = df_R4.apply(lambda row: row['sequence'].split('RE')[1].split('YA')[0] if row['CDR2'] == 1 and 'RE' in row['sequence'] and 'YA' in row['sequence'] else row['CDR2'], axis=1)\n",
    "\n",
    "df_R4.loc[:, 'CDR3'] = df_R4.apply(lambda row: row['sequence'].split('YCA')[1].split('YWG')[0] if row['CDR3'] == 1 and 'YCA' in row['sequence'] and 'YWG' in row['sequence'] else row['CDR3'], axis=1)\n",
    "\n",
    "# Third step: Modify 'CDR3' for rows where 'CDR3' is 1, try finding the sequence between 'CA' and 'YW'\n",
    "df_R4.loc[:, 'CDR3'] = df_R4.apply(lambda row: row['sequence'].split('YCA')[1].split('YW')[0] if row['CDR3'] == 1 and 'YCA' in row['sequence'] and 'YW' in row['sequence'] else row['CDR3'], axis=1)\n",
    "\n",
    "# Fourth step: Modify 'CDR1' for rows where 'CDR1' is still 1, extract sequence positions 19 to 27\n",
    "df_R4.loc[:, 'CDR1'] = df_R4.apply(lambda row: row['sequence'][19:27] if row['CDR1'] == 1 else row['CDR1'], axis=1)\n",
    "\n",
    "# Fifth step: Modify 'CDR2' for rows where 'CDR2' is still 1, extract sequence positions 39 to 52\n",
    "df_R4.loc[:, 'CDR2'] = df_R4.apply(lambda row: row['sequence'][39:52] if row['CDR2'] == 1 else row['CDR2'], axis=1)\n",
    "\n",
    "##ENR\n",
    "# First step: Modify 'CDR1' for rows where 'CDR1' is 1, try finding the sequence between 'SG' and 'MG'\n",
    "NBEnrichment.loc[:, 'CDR1'] = NBEnrichment.apply(lambda row: row['sequence'].split('SG')[1].split('MG')[0] if row['CDR1'] == 1 and 'SG' in row['sequence'] and 'MG' in row['sequence'] else row['CDR1'], axis=1)\n",
    "\n",
    "# Second step: Modify 'CDR2' for rows where 'CDR2' is 1, try finding the sequence between 'RE' and 'YA'\n",
    "NBEnrichment.loc[:, 'CDR2'] = NBEnrichment.apply(lambda row: row['sequence'].split('RE')[1].split('YA')[0] if row['CDR2'] == 1 and 'RE' in row['sequence'] and 'YA' in row['sequence'] else row['CDR2'], axis=1)\n",
    "\n",
    "NBEnrichment.loc[:, 'CDR3'] = NBEnrichment.apply(lambda row: row['sequence'].split('YCA')[1].split('YWG')[0] if row['CDR3'] == 1 and 'YCA' in row['sequence'] and 'YWG' in row['sequence'] else row['CDR3'], axis=1)\n",
    "\n",
    "# Third step: Modify 'CDR3' for rows where 'CDR3' is 1, try finding the sequence between 'CA' and 'YW'\n",
    "NBEnrichment.loc[:, 'CDR3'] = NBEnrichment.apply(lambda row: row['sequence'].split('YCA')[1].split('YW')[0] if row['CDR3'] == 1 and 'YCA' in row['sequence'] and 'YW' in row['sequence'] else row['CDR3'], axis=1)\n",
    "\n",
    "# Fourth step: Modify 'CDR1' for rows where 'CDR1' is still 1, extract sequence positions 19 to 27\n",
    "NBEnrichment.loc[:, 'CDR1'] = NBEnrichment.apply(lambda row: row['sequence'][19:27] if row['CDR1'] == 1 else row['CDR1'], axis=1)\n",
    "\n",
    "# Fifth step: Modify 'CDR2' for rows where 'CDR2' is still 1, extract sequence positions 39 to 52\n",
    "NBEnrichment.loc[:, 'CDR2'] = NBEnrichment.apply(lambda row: row['sequence'][39:52] if row['CDR2'] == 1 else row['CDR2'], axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2641a7a2-4c16-48e0-90ac-ba6c73803954",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Assign the length of 'CDR3' using .loc\n",
    "df_R2.loc[:, 'CDR3_length'] = df_R2['CDR3'].apply(lambda x: len(x) if isinstance(x, str) else 0)\n",
    "df_R3.loc[:, 'CDR3_length'] = df_R3['CDR3'].apply(lambda x: len(x) if isinstance(x, str) else 0)\n",
    "df_R4.loc[:, 'CDR3_length'] = df_R4['CDR3'].apply(lambda x: len(x) if isinstance(x, str) else 0)\n",
    "NBEnrichment.loc[:, 'CDR3_length'] = NBEnrichment['CDR3'].apply(lambda x: len(x) if isinstance(x, str) else 0)\n",
    "# Rename the 'size' column to 'Number of Reads' without inplace=True\n",
    "df_R2 = df_R2.rename(columns={'size': 'Number of Reads'})\n",
    "df_R3 = df_R3.rename(columns={'size': 'Number of Reads'})\n",
    "df_R4 = df_R4.rename(columns={'size': 'Number of Reads'})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b474bddf-8e6a-4857-b420-c972926c0168",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>size_R2</th>\n",
       "      <th>size_R3</th>\n",
       "      <th>size_R4</th>\n",
       "      <th>Enrichment R3/R2</th>\n",
       "      <th>Enrichment R4/R3</th>\n",
       "      <th>Enrichment R4/R2</th>\n",
       "      <th>sequence_length</th>\n",
       "      <th>CDR1</th>\n",
       "      <th>CDR2</th>\n",
       "      <th>CDR3</th>\n",
       "      <th>CDR3_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>QVQLQESGGGLVQAGGSLRLSCAASGSIFTDTYMGWYRQAPGKERE...</td>\n",
       "      <td>117.0</td>\n",
       "      <td>792.0</td>\n",
       "      <td>872.0</td>\n",
       "      <td>6.769231</td>\n",
       "      <td>1.101010</td>\n",
       "      <td>7.452991</td>\n",
       "      <td>122</td>\n",
       "      <td>SIFTDTY</td>\n",
       "      <td>LVATITPGSSTN</td>\n",
       "      <td>AYTYVENQGYYYFA</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>QVQLQESGGGLVQAGGSLRLSCAASGYIFGGGPMGWYRQAPGKERE...</td>\n",
       "      <td>281.0</td>\n",
       "      <td>1310.0</td>\n",
       "      <td>793.0</td>\n",
       "      <td>4.661922</td>\n",
       "      <td>0.605344</td>\n",
       "      <td>2.822064</td>\n",
       "      <td>126</td>\n",
       "      <td>YIFGGGP</td>\n",
       "      <td>FVATIDPGGITY</td>\n",
       "      <td>AEYTEVDFGSNYWSYGHF</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>QVQLQESGGGLVQAGGSLRLSCAASGTIFHGQFMGWYRQAPGKERE...</td>\n",
       "      <td>312.0</td>\n",
       "      <td>3453.0</td>\n",
       "      <td>682.0</td>\n",
       "      <td>11.067308</td>\n",
       "      <td>0.197509</td>\n",
       "      <td>2.185897</td>\n",
       "      <td>118</td>\n",
       "      <td>TIFHGQF</td>\n",
       "      <td>LVATIGSGATTY</td>\n",
       "      <td>VDRGDTGPYP</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>QVQLQESGGGLVQAGDSLRLSCAASGSIFSYYAMGWYRQAPGKERE...</td>\n",
       "      <td>527.0</td>\n",
       "      <td>1838.0</td>\n",
       "      <td>655.0</td>\n",
       "      <td>3.487666</td>\n",
       "      <td>0.356366</td>\n",
       "      <td>1.242884</td>\n",
       "      <td>118</td>\n",
       "      <td>SIFSYYA</td>\n",
       "      <td>FVAGIANGAITN</td>\n",
       "      <td>AYEIDWPYFG</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>QVQLQESGGGLVQAGGSLRLSCAASGYIFVGTYMGWYRQAPGKERE...</td>\n",
       "      <td>69.0</td>\n",
       "      <td>573.0</td>\n",
       "      <td>638.0</td>\n",
       "      <td>8.304348</td>\n",
       "      <td>1.113438</td>\n",
       "      <td>9.246377</td>\n",
       "      <td>122</td>\n",
       "      <td>YIFVGTY</td>\n",
       "      <td>FVAGIASGSSTY</td>\n",
       "      <td>VFYGDQDDSNILYT</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30622</th>\n",
       "      <td>QVQLQESGGGLVQAGGSLRLSCAASGNIFRPGTMGWYRQAPGKERE...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>118</td>\n",
       "      <td>NIFRPGT</td>\n",
       "      <td>FVASISLGATTY</td>\n",
       "      <td>AEGDEFVDLY</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30623</th>\n",
       "      <td>QVQLQESGGGLVQAGGSLRLSCAASGTISRLPSMGWYRQAPGKERE...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>122</td>\n",
       "      <td>TISRLPS</td>\n",
       "      <td>LVAGIALGSTTN</td>\n",
       "      <td>VGYDYNTDIKYAYY</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14789</th>\n",
       "      <td>QVQLQESGGGLVQAGGSLRLSCAASGSISTGDGMGWYRQAPGKERE...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>124</td>\n",
       "      <td>SISTGDG</td>\n",
       "      <td>FVAAISTGAITY</td>\n",
       "      <td>AYGVASQEAVYYLD</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30625</th>\n",
       "      <td>QVQLQESGGGLVQAGGSLRLSCAASGYISSSQSMGWYRQAPGKERE...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>126</td>\n",
       "      <td>YISSSQS</td>\n",
       "      <td>LVAAIATGTTTY</td>\n",
       "      <td>APQSDYGRDLQAYWWYYL</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30240</th>\n",
       "      <td>QVQLQESGGGLVQAGGSLRLSCAASGTISTAGRMGWYRQAPGKERE...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>122</td>\n",
       "      <td>TISTAGR</td>\n",
       "      <td>LVAGINYGAITY</td>\n",
       "      <td>AVHSPDNLVLGGLD</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60480 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                sequence  size_R2  size_R3  \\\n",
       "40     QVQLQESGGGLVQAGGSLRLSCAASGSIFTDTYMGWYRQAPGKERE...    117.0    792.0   \n",
       "7      QVQLQESGGGLVQAGGSLRLSCAASGYIFGGGPMGWYRQAPGKERE...    281.0   1310.0   \n",
       "5      QVQLQESGGGLVQAGGSLRLSCAASGTIFHGQFMGWYRQAPGKERE...    312.0   3453.0   \n",
       "1      QVQLQESGGGLVQAGDSLRLSCAASGSIFSYYAMGWYRQAPGKERE...    527.0   1838.0   \n",
       "88     QVQLQESGGGLVQAGGSLRLSCAASGYIFVGTYMGWYRQAPGKERE...     69.0    573.0   \n",
       "...                                                  ...      ...      ...   \n",
       "30622  QVQLQESGGGLVQAGGSLRLSCAASGNIFRPGTMGWYRQAPGKERE...      1.0      2.0   \n",
       "30623  QVQLQESGGGLVQAGGSLRLSCAASGTISRLPSMGWYRQAPGKERE...      1.0      2.0   \n",
       "14789  QVQLQESGGGLVQAGGSLRLSCAASGSISTGDGMGWYRQAPGKERE...      1.0      0.0   \n",
       "30625  QVQLQESGGGLVQAGGSLRLSCAASGYISSSQSMGWYRQAPGKERE...      1.0      2.0   \n",
       "30240  QVQLQESGGGLVQAGGSLRLSCAASGTISTAGRMGWYRQAPGKERE...      1.0      2.0   \n",
       "\n",
       "       size_R4  Enrichment R3/R2  Enrichment R4/R3  Enrichment R4/R2  \\\n",
       "40       872.0          6.769231          1.101010          7.452991   \n",
       "7        793.0          4.661922          0.605344          2.822064   \n",
       "5        682.0         11.067308          0.197509          2.185897   \n",
       "1        655.0          3.487666          0.356366          1.242884   \n",
       "88       638.0          8.304348          1.113438          9.246377   \n",
       "...        ...               ...               ...               ...   \n",
       "30622      0.0          2.000000          0.000000          0.000000   \n",
       "30623      0.0          2.000000          0.000000          0.000000   \n",
       "14789      0.0          0.000000               NaN          0.000000   \n",
       "30625      0.0          2.000000          0.000000          0.000000   \n",
       "30240      0.0          2.000000          0.000000          0.000000   \n",
       "\n",
       "       sequence_length     CDR1          CDR2                CDR3  CDR3_length  \n",
       "40                 122  SIFTDTY  LVATITPGSSTN      AYTYVENQGYYYFA           14  \n",
       "7                  126  YIFGGGP  FVATIDPGGITY  AEYTEVDFGSNYWSYGHF           18  \n",
       "5                  118  TIFHGQF  LVATIGSGATTY          VDRGDTGPYP           10  \n",
       "1                  118  SIFSYYA  FVAGIANGAITN          AYEIDWPYFG           10  \n",
       "88                 122  YIFVGTY  FVAGIASGSSTY      VFYGDQDDSNILYT           14  \n",
       "...                ...      ...           ...                 ...          ...  \n",
       "30622              118  NIFRPGT  FVASISLGATTY          AEGDEFVDLY           10  \n",
       "30623              122  TISRLPS  LVAGIALGSTTN      VGYDYNTDIKYAYY           14  \n",
       "14789              124  SISTGDG  FVAAISTGAITY      AYGVASQEAVYYLD           14  \n",
       "30625              126  YISSSQS  LVAAIATGTTTY  APQSDYGRDLQAYWWYYL           18  \n",
       "30240              122  TISTAGR  LVAGINYGAITY      AVHSPDNLVLGGLD           14  \n",
       "\n",
       "[60480 rows x 12 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NBEnrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "43f6f01d-cf0b-48d8-92c0-59e8653fe477",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence_length</th>\n",
       "      <th>sequence</th>\n",
       "      <th>Number of Reads</th>\n",
       "      <th>CDR1</th>\n",
       "      <th>CDR2</th>\n",
       "      <th>CDR3</th>\n",
       "      <th>CDR3_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7432</th>\n",
       "      <td>122</td>\n",
       "      <td>QVQLQESGGGLVQAGGSLRLSCAASGSIFTDTYMGWYRQAPGKERE...</td>\n",
       "      <td>872</td>\n",
       "      <td>SIFTDTY</td>\n",
       "      <td>LVATITPGSSTN</td>\n",
       "      <td>AYTYVENQGYYYFA</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14951</th>\n",
       "      <td>126</td>\n",
       "      <td>QVQLQESGGGLVQAGGSLRLSCAASGYIFGGGPMGWYRQAPGKERE...</td>\n",
       "      <td>793</td>\n",
       "      <td>YIFGGGP</td>\n",
       "      <td>FVATIDPGGITY</td>\n",
       "      <td>AEYTEVDFGSNYWSYGHF</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2696</th>\n",
       "      <td>118</td>\n",
       "      <td>QVQLQESGGGLVQAGGSLRLSCAASGTIFHGQFMGWYRQAPGKERE...</td>\n",
       "      <td>682</td>\n",
       "      <td>TIFHGQF</td>\n",
       "      <td>LVATIGSGATTY</td>\n",
       "      <td>VDRGDTGPYP</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>834</th>\n",
       "      <td>118</td>\n",
       "      <td>QVQLQESGGGLVQAGDSLRLSCAASGSIFSYYAMGWYRQAPGKERE...</td>\n",
       "      <td>655</td>\n",
       "      <td>SIFSYYA</td>\n",
       "      <td>FVAGIANGAITN</td>\n",
       "      <td>AYEIDWPYFG</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10768</th>\n",
       "      <td>122</td>\n",
       "      <td>QVQLQESGGGLVQAGGSLRLSCAASGYIFVGTYMGWYRQAPGKERE...</td>\n",
       "      <td>638</td>\n",
       "      <td>YIFVGTY</td>\n",
       "      <td>FVAGIASGSSTY</td>\n",
       "      <td>VFYGDQDDSNILYT</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5963</th>\n",
       "      <td>122</td>\n",
       "      <td>QVQLQESGGGLVQAGGGLRLSCAASGSIFNDVHMGWYRQAPGKERE...</td>\n",
       "      <td>1</td>\n",
       "      <td>SIFNDVH</td>\n",
       "      <td>FVAGIANGGSTY</td>\n",
       "      <td>VEAYVDRRFAYYYN</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5964</th>\n",
       "      <td>122</td>\n",
       "      <td>QVQLQESGGGLVQAGGGLRLSCAASGSIFNDVHMGWYRQAPGKERE...</td>\n",
       "      <td>1</td>\n",
       "      <td>SIFNDVH</td>\n",
       "      <td>LVAGISNGGSTY</td>\n",
       "      <td>AYGGSRLFIRYWFA</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5965</th>\n",
       "      <td>122</td>\n",
       "      <td>QVQLQESGGGLVQAGGGLRLSCAASGSIFSYYAMGWYRQAPGKERE...</td>\n",
       "      <td>1</td>\n",
       "      <td>SIFSYYA</td>\n",
       "      <td>FVAGIANGAITN</td>\n",
       "      <td>ATDLWSFRDEWAYD</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5966</th>\n",
       "      <td>122</td>\n",
       "      <td>QVQLQESGGGLVQAGGGLRLSCAASGSIFTDTYMGWYRQAPGKERE...</td>\n",
       "      <td>1</td>\n",
       "      <td>SIFTDTY</td>\n",
       "      <td>FVAAIDNGANTY</td>\n",
       "      <td>ASYPLHYLDTWWFI</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16005</th>\n",
       "      <td>131</td>\n",
       "      <td>QVQLQESGGGLVQPGGSLRLSCTASGFTLDYYNIGWFRQAPGKERE...</td>\n",
       "      <td>1</td>\n",
       "      <td>LSCTASGF</td>\n",
       "      <td>RVSCISSSGGRTN</td>\n",
       "      <td>AEYTEVDFGSNYWSYGHF</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16006 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       sequence_length                                           sequence  \\\n",
       "7432               122  QVQLQESGGGLVQAGGSLRLSCAASGSIFTDTYMGWYRQAPGKERE...   \n",
       "14951              126  QVQLQESGGGLVQAGGSLRLSCAASGYIFGGGPMGWYRQAPGKERE...   \n",
       "2696               118  QVQLQESGGGLVQAGGSLRLSCAASGTIFHGQFMGWYRQAPGKERE...   \n",
       "834                118  QVQLQESGGGLVQAGDSLRLSCAASGSIFSYYAMGWYRQAPGKERE...   \n",
       "10768              122  QVQLQESGGGLVQAGGSLRLSCAASGYIFVGTYMGWYRQAPGKERE...   \n",
       "...                ...                                                ...   \n",
       "5963               122  QVQLQESGGGLVQAGGGLRLSCAASGSIFNDVHMGWYRQAPGKERE...   \n",
       "5964               122  QVQLQESGGGLVQAGGGLRLSCAASGSIFNDVHMGWYRQAPGKERE...   \n",
       "5965               122  QVQLQESGGGLVQAGGGLRLSCAASGSIFSYYAMGWYRQAPGKERE...   \n",
       "5966               122  QVQLQESGGGLVQAGGGLRLSCAASGSIFTDTYMGWYRQAPGKERE...   \n",
       "16005              131  QVQLQESGGGLVQPGGSLRLSCTASGFTLDYYNIGWFRQAPGKERE...   \n",
       "\n",
       "       Number of Reads      CDR1           CDR2                CDR3  \\\n",
       "7432               872   SIFTDTY   LVATITPGSSTN      AYTYVENQGYYYFA   \n",
       "14951              793   YIFGGGP   FVATIDPGGITY  AEYTEVDFGSNYWSYGHF   \n",
       "2696               682   TIFHGQF   LVATIGSGATTY          VDRGDTGPYP   \n",
       "834                655   SIFSYYA   FVAGIANGAITN          AYEIDWPYFG   \n",
       "10768              638   YIFVGTY   FVAGIASGSSTY      VFYGDQDDSNILYT   \n",
       "...                ...       ...            ...                 ...   \n",
       "5963                 1   SIFNDVH   FVAGIANGGSTY      VEAYVDRRFAYYYN   \n",
       "5964                 1   SIFNDVH   LVAGISNGGSTY      AYGGSRLFIRYWFA   \n",
       "5965                 1   SIFSYYA   FVAGIANGAITN      ATDLWSFRDEWAYD   \n",
       "5966                 1   SIFTDTY   FVAAIDNGANTY      ASYPLHYLDTWWFI   \n",
       "16005                1  LSCTASGF  RVSCISSSGGRTN  AEYTEVDFGSNYWSYGHF   \n",
       "\n",
       "       CDR3_length  \n",
       "7432            14  \n",
       "14951           18  \n",
       "2696            10  \n",
       "834             10  \n",
       "10768           14  \n",
       "...            ...  \n",
       "5963            14  \n",
       "5964            14  \n",
       "5965            14  \n",
       "5966            14  \n",
       "16005           18  \n",
       "\n",
       "[16006 rows x 7 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_R4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "9601393c-b77e-4a42-8080-3451a464e21c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_R4.to_csv('./Results/HARP-NB-Round4-ALL.csv', index=False)\n",
    "df_R4 = df_R4.loc[df_R4['CDR3'] != 1] #(filtering out CDR3s with error)\n",
    "df_R4.to_csv('./Results/HARP-NB-Round4-Clean.csv', index=False)\n",
    "NBEnrichment.to_csv(\"./Results/HARP-NB-Enrichment-final.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "428464c0-33ae-45f7-8737-a93b4f35ff75",
   "metadata": {},
   "source": [
    "## CDR3 GROUPING"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "2245cb8b-28df-460e-a6b7-72f69d5057ab",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CDR3</th>\n",
       "      <th>Number of Reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AYEIDWPYFG</td>\n",
       "      <td>2006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>VDRGDTGPYP</td>\n",
       "      <td>1961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AYTYVENQGYYYFA</td>\n",
       "      <td>1899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AEYTEVDFGSNYWSYGHF</td>\n",
       "      <td>1590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ADVAIYYWLL</td>\n",
       "      <td>1134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1273</th>\n",
       "      <td>AYWLGDVLYQ</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1274</th>\n",
       "      <td>AYYDTSSYLVYGHI</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1275</th>\n",
       "      <td>AYYHGPDRLV</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1276</th>\n",
       "      <td>AYYPHDHTVGFEHD</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1277</th>\n",
       "      <td>DWTVNNVGYG</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1278 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    CDR3  Number of Reads\n",
       "0             AYEIDWPYFG             2006\n",
       "1             VDRGDTGPYP             1961\n",
       "2         AYTYVENQGYYYFA             1899\n",
       "3     AEYTEVDFGSNYWSYGHF             1590\n",
       "4             ADVAIYYWLL             1134\n",
       "...                  ...              ...\n",
       "1273          AYWLGDVLYQ                1\n",
       "1274      AYYDTSSYLVYGHI                1\n",
       "1275          AYYHGPDRLV                1\n",
       "1276      AYYPHDHTVGFEHD                1\n",
       "1277          DWTVNNVGYG                1\n",
       "\n",
       "[1278 rows x 2 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Group by 'CDR3' and sum the product of 'Number of Reads'\n",
    "df_R4_CDR3 = df_R4.groupby('CDR3').agg({'Number of Reads': 'sum'}).reset_index()\n",
    "df_R2_CDR3 = df_R2.groupby('CDR3').agg({'Number of Reads': 'sum'}).reset_index()\n",
    "\n",
    "df_R2 = df_R2.loc[df_R2['CDR3'] != 1]\n",
    "df_R2_CDR3.sort_values(by= \"Number of Reads\", ascending= False, inplace=True)\n",
    "df_R2_CDR3.reset_index(drop=True, inplace=True)\n",
    "df_R2_CDR3.to_csv('./Results/HARP-NB-Round2-CDR3.csv', index=False)\n",
    "\n",
    "df_R4_CDR3.sort_values(by= \"Number of Reads\", ascending= False, inplace=True)\n",
    "df_R4_CDR3.reset_index(drop=True, inplace=True)\n",
    "df_R4_CDR3.to_csv('./Results/HARP-NB-Round4-CDR3.csv', index=False)\n",
    "df_R4_CDR3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "9686f189-2d50-4495-82e9-99cb562f3b8d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NGS Enrichment Analysis done sucsesfully.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CDR3</th>\n",
       "      <th>R2</th>\n",
       "      <th>R4</th>\n",
       "      <th>Enrichment Ratio</th>\n",
       "      <th>CDR3_length</th>\n",
       "      <th>Rand</th>\n",
       "      <th>Only Random All</th>\n",
       "      <th>Random Region Middle</th>\n",
       "      <th>CDR3_length-rmid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AYEIDWPYFG</td>\n",
       "      <td>1100.0</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>1.823636</td>\n",
       "      <td>10</td>\n",
       "      <td>YEIDWPYFG</td>\n",
       "      <td>YEIDWPYG</td>\n",
       "      <td>YEIDWPY</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>VDRGDTGPYP</td>\n",
       "      <td>789.0</td>\n",
       "      <td>1961.0</td>\n",
       "      <td>2.485425</td>\n",
       "      <td>10</td>\n",
       "      <td>DRGDTGPYP</td>\n",
       "      <td>DRGDTGPP</td>\n",
       "      <td>DRGDTGP</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>AYTYVENQGYYYFA</td>\n",
       "      <td>301.0</td>\n",
       "      <td>1899.0</td>\n",
       "      <td>6.308970</td>\n",
       "      <td>14</td>\n",
       "      <td>YTYVENQGYYYFA</td>\n",
       "      <td>YTYVENQGYYYA</td>\n",
       "      <td>YTYVENQGYYY</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AEYTEVDFGSNYWSYGHF</td>\n",
       "      <td>534.0</td>\n",
       "      <td>1590.0</td>\n",
       "      <td>2.977528</td>\n",
       "      <td>18</td>\n",
       "      <td>EYTEVDFGSNYWSYGHF</td>\n",
       "      <td>EYTEVDFGSNYWSYGF</td>\n",
       "      <td>EYTEVDFGSNYWSYG</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>ADVAIYYWLL</td>\n",
       "      <td>126.0</td>\n",
       "      <td>1134.0</td>\n",
       "      <td>9.000000</td>\n",
       "      <td>10</td>\n",
       "      <td>DVAIYYWLL</td>\n",
       "      <td>DVAIYYWL</td>\n",
       "      <td>DVAIYYW</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1435</th>\n",
       "      <td>VSYDLDLYGLYPFE</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>14</td>\n",
       "      <td>SYDLDLYGLYPFE</td>\n",
       "      <td>SYDLDLYGLYPE</td>\n",
       "      <td>SYDLDLYGLYP</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1434</th>\n",
       "      <td>VSYDLDLYGLSPYE</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>14</td>\n",
       "      <td>SYDLDLYGLSPYE</td>\n",
       "      <td>SYDLDLYGLSPE</td>\n",
       "      <td>SYDLDLYGLSP</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1433</th>\n",
       "      <td>ATDYVLRSLN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10</td>\n",
       "      <td>TDYVLRSLN</td>\n",
       "      <td>TDYVLRSN</td>\n",
       "      <td>TDYVLRS</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1432</th>\n",
       "      <td>ADGHNAYWFP</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10</td>\n",
       "      <td>DGHNAYWFP</td>\n",
       "      <td>DGHNAYWP</td>\n",
       "      <td>DGHNAYW</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1367</th>\n",
       "      <td>ANESTDEYFR</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10</td>\n",
       "      <td>NESTDEYFR</td>\n",
       "      <td>NESTDEYR</td>\n",
       "      <td>NESTDEY</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2734 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    CDR3      R2      R4  Enrichment Ratio  CDR3_length  \\\n",
       "2             AYEIDWPYFG  1100.0  2006.0          1.823636           10   \n",
       "3             VDRGDTGPYP   789.0  1961.0          2.485425           10   \n",
       "17        AYTYVENQGYYYFA   301.0  1899.0          6.308970           14   \n",
       "8     AEYTEVDFGSNYWSYGHF   534.0  1590.0          2.977528           18   \n",
       "92            ADVAIYYWLL   126.0  1134.0          9.000000           10   \n",
       "...                  ...     ...     ...               ...          ...   \n",
       "1435      VSYDLDLYGLYPFE     1.0     0.0          0.000000           14   \n",
       "1434      VSYDLDLYGLSPYE     1.0     0.0          0.000000           14   \n",
       "1433          ATDYVLRSLN     1.0     0.0          0.000000           10   \n",
       "1432          ADGHNAYWFP     1.0     0.0          0.000000           10   \n",
       "1367          ANESTDEYFR     2.0     0.0          0.000000           10   \n",
       "\n",
       "                   Rand   Only Random All Random Region Middle  \\\n",
       "2             YEIDWPYFG          YEIDWPYG              YEIDWPY   \n",
       "3             DRGDTGPYP          DRGDTGPP              DRGDTGP   \n",
       "17        YTYVENQGYYYFA      YTYVENQGYYYA          YTYVENQGYYY   \n",
       "8     EYTEVDFGSNYWSYGHF  EYTEVDFGSNYWSYGF      EYTEVDFGSNYWSYG   \n",
       "92            DVAIYYWLL          DVAIYYWL              DVAIYYW   \n",
       "...                 ...               ...                  ...   \n",
       "1435      SYDLDLYGLYPFE      SYDLDLYGLYPE          SYDLDLYGLYP   \n",
       "1434      SYDLDLYGLSPYE      SYDLDLYGLSPE          SYDLDLYGLSP   \n",
       "1433          TDYVLRSLN          TDYVLRSN              TDYVLRS   \n",
       "1432          DGHNAYWFP          DGHNAYWP              DGHNAYW   \n",
       "1367          NESTDEYFR          NESTDEYR              NESTDEY   \n",
       "\n",
       "      CDR3_length-rmid  \n",
       "2                    7  \n",
       "3                    7  \n",
       "17                  11  \n",
       "8                   15  \n",
       "92                   7  \n",
       "...                ...  \n",
       "1435                11  \n",
       "1434                11  \n",
       "1433                 7  \n",
       "1432                 7  \n",
       "1367                 7  \n",
       "\n",
       "[2734 rows x 9 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## CDR3 Enrichment\n",
    "# Enrichment\n",
    "\n",
    "def Enrichment (df_naive , df_sorted, outfile):\n",
    "    df1= df_naive\n",
    "    df2= df_sorted\n",
    "    \n",
    "    # Merge dataframes with left join\n",
    "    merged_df = pd.merge(df1, df2, on='CDR3', how='outer')\n",
    "    # Replace NaN values with 1 in the 'size_x' column\n",
    "    merged_df['Number of Reads_x'] = merged_df['Number of Reads_x'].fillna(1)\n",
    "\n",
    "    # Replace NaN values with 1 in the 'size_x' column\n",
    "    merged_df['Number of Reads_y'] = merged_df['Number of Reads_y'].fillna(0)\n",
    "    #rename\n",
    "    merged_df.rename(columns = {'Number of Reads_x':'R2' , 'Number of Reads_y':'R4'}, inplace = True)\n",
    "    #enrichment ratio\n",
    "    merged_df [\"Enrichment Ratio\"] = merged_df [\"R4\"] / merged_df [\"R2\"]\n",
    "\n",
    "    #save to file\n",
    "    merged_df.sort_values(by= \"R4\", ascending= False, inplace=True)\n",
    "    merged_df['CDR3_length'] = merged_df['CDR3'].astype(str).apply(len)\n",
    "    #ADD Random seq\n",
    "    merged_df['Rand'] = merged_df['CDR3'].str.slice(1)\n",
    "    # Remove rows with NaN in the 'Rand' column\n",
    "    merged_df.dropna(subset=['Rand'], inplace=True)\n",
    "\n",
    "    # Function to remove the character before the last one from each sequence\n",
    "    def remove_char_before_last(seq):\n",
    "        if len(seq) > 1:\n",
    "            return seq[:-2] + seq[-1]  # Keep everything except the second-to-last character and append the last character\n",
    "        else:\n",
    "            return seq  # If the sequence has only one character, leave it as it is\n",
    "    merged_df['Only Random All'] = merged_df['Rand'].apply(remove_char_before_last)\n",
    "    merged_df [\"Random Region Middle\"] = merged_df [\"Rand\"].astype(str).str[:-2]\n",
    "    merged_df.loc[:, 'CDR3_length-rmid'] = merged_df['Random Region Middle'].apply(lambda x: len(x) if isinstance(x, str) else 0)\n",
    "    outdf = merged_df\n",
    "    outdf.to_csv(outfile, index=True)\n",
    "    print(\"NGS Enrichment Analysis done sucsesfully.\")\n",
    "    return outdf\n",
    "    \n",
    "    \n",
    "CDR3_Enrichment = Enrichment (df_R2_CDR3, df_R4_CDR3,'./Results/CDR3_Enrichment.csv')\n",
    "CDR3_Enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "05b07498-e3b2-4115-b6fa-6f2c64696bfc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of sequences containing 'ENQ': 16\n"
     ]
    }
   ],
   "source": [
    "##Find sequensces with ENQ \n",
    "\n",
    "df_R4_CDR3 = pd.read_csv(\"./Results/HARP-NB-Round4-CDR3.csv\")  \n",
    "\n",
    "def count_sequences_with_enq(df, column_name=\"CDR3\"):\n",
    "    \"\"\"\n",
    "    Counts the number of sequences in the specified column that contain the substring \"ENQ\".\n",
    "    \n",
    "    Parameters:\n",
    "        df (pd.DataFrame): Input dataframe containing sequence data.\n",
    "        column_name (str): The name of the column to check. Default is \"CDR3\".\n",
    "    \n",
    "    Returns:\n",
    "        int: Count of sequences containing \"ENQ\"\n",
    "    \"\"\"\n",
    "    return df[column_name].str.contains(\"ENQ\", na=False).sum()\n",
    "\n",
    "\n",
    "\n",
    "count = count_sequences_with_enq(df_R4_CDR3)\n",
    "print(f\"Number of sequences containing 'ENQ': {count}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "ea38210d-d6eb-42e2-a700-8faa3c5d38f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CDR3</th>\n",
       "      <th>Number of Reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AYTYVENQGYYYFA</td>\n",
       "      <td>1899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>614</th>\n",
       "      <td>AYTYVENQGYYYLA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>639</th>\n",
       "      <td>AYTYVENQGYYYFV</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>712</th>\n",
       "      <td>AYTYVENQDYYYFA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>714</th>\n",
       "      <td>AYTYVENQGYYEFP</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>755</th>\n",
       "      <td>VYTYVENQGYYYFA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1193</th>\n",
       "      <td>AYSYVENQGYYYFA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1197</th>\n",
       "      <td>AYTNVENQGYYYFA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1212</th>\n",
       "      <td>AYAYVENQGYYYFA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1248</th>\n",
       "      <td>SYTYVENQGYYYFA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1256</th>\n",
       "      <td>AYTYVENQVYYYFD</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1261</th>\n",
       "      <td>AYTYVENQGYFDYYYI</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1262</th>\n",
       "      <td>AYTYVENQGYHYFA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1263</th>\n",
       "      <td>AYTYVENQGYSYFA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1264</th>\n",
       "      <td>AYTYVENQGYYA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1265</th>\n",
       "      <td>AYTYVENQGYYYFD</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  CDR3  Number of Reads\n",
       "2       AYTYVENQGYYYFA             1899\n",
       "614     AYTYVENQGYYYLA                2\n",
       "639     AYTYVENQGYYYFV                2\n",
       "712     AYTYVENQDYYYFA                2\n",
       "714     AYTYVENQGYYEFP                2\n",
       "755     VYTYVENQGYYYFA                1\n",
       "1193    AYSYVENQGYYYFA                1\n",
       "1197    AYTNVENQGYYYFA                1\n",
       "1212    AYAYVENQGYYYFA                1\n",
       "1248    SYTYVENQGYYYFA                1\n",
       "1256    AYTYVENQVYYYFD                1\n",
       "1261  AYTYVENQGYFDYYYI                1\n",
       "1262    AYTYVENQGYHYFA                1\n",
       "1263    AYTYVENQGYSYFA                1\n",
       "1264      AYTYVENQGYYA                1\n",
       "1265    AYTYVENQGYYYFD                1"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_enq = df_R4_CDR3[df_R4_CDR3[\"CDR3\"].str.contains(\"ENQ\")]\n",
    "df_enq"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bio",
   "language": "python",
   "name": "bio"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
